include/envoy/ratelimit/ratelimit.h

```diff
@@ -39,21 +45,62 @@ struct Descriptor {
 };
 
 /**
+ * A single token bucket. See token_bucket.proto.
+ */
+struct TokenBucket {
+  uint32_t max_tokens_;
+  uint32_t tokens_per_fill_;
+  absl::Duration fill_interval_;
+};
+
+/**
+ * A token bucket operational state.
+ */
+struct TokenState {
+  mutable std::atomic<uint32_t> tokens_;
+  MonotonicTime fill_time_;
+};
+
+/**
+ * A single rate limit request descriptor. See ratelimit.proto.
+ * The descriptor entries constitute the primary key for local descriptors.
+ */
+struct LocalDescriptor {
+  std::vector<DescriptorEntry> entries_;
+  TokenBucket token_bucket_ = {};
+  TokenState* token_state_ = nullptr;
```

Comment:
Idea: Is it possible to have an impl struct that derives from LocalDescriptor with the token state so we don't need to expose it in the public headers? Not sure if that is possible or not, but seems like it might be.

Reply:
Yeah, it's possible with heterogeneous absl set.